rfcs/20240312-standardize-tan-customcall-tuplecollective-ops.md

```diff
@@ -0,0 +1,377 @@
+# [RFC] Standardize TanOp, CustomCallOp with typed FFI, and tuple-collectives ops
+
+Status: Draft<br/>
+Initial version: 03/12/2024<br/>
+Last updated: 03/12/2024<br/>
+Discussion thread: [GitHub](add PR Link)
+
+## Motivation
+
+Several features have been added to MHLO in the past year, which frameworks want
+to leverage and members of the community have made requests for them as well.
+This includes: `TanOp`, `CustomCallOp` with typed FFI (dictionary), and
+tuple-collective ops (`AllGatherOp`, `AllReduceOp`, `AllToAllOp` with variadic
+operands/results). Some of these features are being used today with various
+workarounds -- unregistered attributes, serializing dictionary as string,
+custom_calls. None of these approaches are stable, so we propose adding these
+ops/features to the StableHLO spec so they can be used safely by the community.
+
+## TanOp
+
+### Rationale
+
+Frameworks and Compilers both want `tan` op.
+Jax has [`jnp.tan`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.tan.html),
+PyTorch has [`torch.tan`](https://pytorch.org/docs/stable/generated/torch.tan.html).
+On Compilers side, XLA has [`mhlo.tan`](https://github.com/tensorflow/mlir-hlo/blob/master/mhlo/IR/hlo_ops.td#L633).
+StableHLO doesn't support `tan` op, but there is an open ticket which requests
+adding this feature in
+[#1358](https://github.com/openxla/stablehlo/issues/1358)
+
+### Proposed Specification
```

Comment:
How much effort would it be to put the changes to existing spec's in bold? This feedback applies mostly to custom call and tuple collectives since Tan is entirely new.

Reply:
reordered the RFC: moved spec changes to spec.md to show diff vs existing spec. added link to rich tests if someone whats to see full spec.